flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -357,11 +359,26 @@ public void resumeConsumption() throws IOException {
         checkState(!isReleased.get(), "Channel released.");
         checkPartitionRequestQueueInitialized();
 
+        if (initialCredit == 0) {
+            unannouncedCredit.set(0);
```

Comment:
Doing this in the onBlockingUpstream method can guarantee that after resumeConsumption the unannouncedCredit is in a clean state. Because the unannouncedCredit may get increased after that if there is any new available floating credit.

Reply:
But how can this floating credit be assigned to this blocked `RemoteInputChannel`? Wouldn't it cause the same deadlock, when floating buffers are assigned to blocked channels and job/task can not make any progress?

It sounds like maybe this should have been handled sooner when trying to increase `unannouncedCredit`?

Reply:
This because RemoteInputChannel#notifyBufferAvailable which increases ```unannouncedCredit``` and floating buffer assignment is not an atomic operation. There is a possibility:
1. the floating buffer is assigned;
2. the floating buffer is released because of receiving an event which blocks the channel;
3. the notifyBufferAvailable is called and the unannouncedCredit is increased.

Do you mean we should make RemoteInputChannel#notifyBufferAvailable and floating buffer assignment an atomic operation?

Reply:
Something sounds wrong here. The race condition that you described above, does it mean that `unannouncedCredit` can be out of sync? That we in reality have released all floating buffers, channel is blocked, but actually `unannouncedCredit > 0`? And it's only fixed after calling `resumeConsumption()`?

Reply:
>>> Something sounds wrong here. The race condition that you described above, does it mean that unannouncedCredit can be out of sync? That we in reality have released all floating buffers, channel is blocked, but actually unannouncedCredit > 0? And it's only fixed after calling resumeConsumption()?

Yes, exactly.

>>> And as I understand it, without your change, this problem doesn't exist, as floating buffers are kept assigned to the blocked channel and the unannouncedCredit (or maybe even assigned AddCredit that might have been sent to the upstream node) are consistent with the reality. Also those assigned floating buffers are not used because channel is blocked, but that is not a big issue, because thanks to the exclusive buffers, other channels can make a progress?

That only happens when the exclusive credit is 0. If the exclusive credit is not 0, the allocated floating buffers will not be released and if the exclusive credit is 0, we release the floating buffers allocated to let other channel use them to avoid deadlock, an extreme case is that we only have 1 floating buffer and no exclusive buffer. At downstream, the unannounced credit will be reset, at the upstream, the available credit is also reset to 0 when resume consumption.

Reply:
Ok, can you at least add a comment explaining why this is set to `0` here?